src/layers/core/icon-layer/icon-layer-vertex.glsl.js

```diff
@@ -50,15 +51,16 @@ vec2 rotate_by_angle(vec2 vertex, float angle) {
 
 void main(void) {
   vec2 iconSize = instanceIconFrames.zw;
-  vec2 iconSize_clipspace = iconSize / viewportSize * 2.0;
   // scale icon height to match instanceSize
   float instanceScale = iconSize.y == 0.0 ? 0.0 : instanceSizes / iconSize.y;
-  vec3 center = project_position(instancePositions);
-  vec2 vertex = (positions / 2.0 + instanceOffsets);
-  vertex = rotate_by_angle(vertex, instanceAngles) * iconSize_clipspace *
-    sizeScale * instanceScale;
+
+  // scale and rotate vertex in "pixel" value and convert back to fraction in clipspace
+  vec2 vertex = positions / 2.0 * iconSize + instanceOffsets;
+  vertex = rotate_by_angle(vertex, instanceAngles) / viewportSize * sizeScale *
+    instanceScale * devicePixelRatio;
```

Comment:
There are other layers using the `viewportSize` uniform to calculate geometry size where they are specified in pixels. In fact I think that's the only reason we have it. I would argue that we should adjust `viewportSize` by `devicePixelRatio` before passing it in, instead of fixing it in individual shaders.

Reply:
> we should adjust viewportSize by devicePixelRatio before passing it in, instead of fixing it in individual shaders.

I have a [PR](https://github.com/uber/deck.gl/pull/902) open on the `project` module. It formalizes and will document the uniforms provided by the project module, which will now include `project_uViewportSize` and `project_uDevicePixelRatio`. The intention is that the latter will be set correctly depending on the application's choice of `useDevicePixelRatio`.

So with these changes, at least layers will not need to inject these uniforms themselves, which is error prone due to the `devicePixelRatio` logic.

I am actually adding a new method `project_pixel_to_clipspace` that will calculate the clipspace offset of a given number of pixels so that it can be added directly to `gl_Position` after `project_to_clipspace`.

Another reason why this function is better than multiplying these directly, is that when using first person viewport, there is no implicit subject distance, so we need to choose a focal plane where the pixel distance is given.

The reason for this it that since we don't consider the `w` coordinate when we add to clipspace, "pixel" values are scaled with distance. Which is normally a desirable effect. 

So by transitioning layers to this function, we will get much cleaner shader code, better control to play with the logic centrally, and solve both DPR and first person view issues.

Reply:
Lightly edited previous post for clarity.